modules/dnn/src/model.cpp

```diff
@@ -637,6 +637,257 @@ void DetectionModel::detect(InputArray frame, CV_OUT std::vector<int>& classIds,
+            versionInt == static_cast<int>(YoloVersion::YOLOv5) ||
+            versionInt == static_cast<int>(YoloVersion::YOLOv6) ||
+            versionInt == static_cast<int>(YoloVersion::YOLOv7) ||
+            versionInt == static_cast<int>(YoloVersion::YOLOv8) ||
+            versionInt == static_cast<int>(YoloVersion::YOLOx)
+            ) {
+            yoloVersion = static_cast<YoloVersion>(versionInt);
+        } else {
+            CV_Error(CV_StsNotImplemented, "Unsupported YOLO version");
+        }
+    }
+
+    YoloVersion getYoloVersion(){
+        return yoloVersion;
+    }
+
+    bool getNetworkType(){
+        return darknet;
+    }
+
+    static void yoloPostProccess(
+        std::vector<Mat>& detections,
+        CV_OUT std::vector<Rect2d>& keep_boxes,
+        CV_OUT std::vector<float>& keep_confidences,
+        CV_OUT std::vector<int>& keep_classIds,
+        const float confThreshold,
+        const float nmsThreshold,
+        YoloVersion yoloVersion,
+        bool darknet){
+
+        // Retrieve
+        std::vector<int> classIds;
+        std::vector<float> confidences;
+        std::vector<Rect2d> boxes;
+
+        if (yoloVersion == YoloVersion::YOLOv8){
+            cv::transposeND(detections[0], {0, 2, 1}, detections[0]);
+        }
+
+        // each row is [cx, cy, w, h, conf_obj, conf_class1, ..., conf_class80]
+        for (auto preds : detections)
+        {
+            if (!darknet)
+                preds = preds.reshape(1, preds.size[1]);
+
+            for (int i = 0; i < preds.rows; ++i)
+            {
+                // filter out non objects
+                float obj_conf = (yoloVersion != YoloVersion::YOLOv8) ? preds.at<float>(i, 4) : 1.0f;
+                if (obj_conf < confThreshold)
+                    continue;
+
+                Mat scores = preds.row(i).colRange((yoloVersion != YoloVersion::YOLOv8) ? 5 : 4, preds.cols);
+                double conf;
+                Point maxLoc;
+                minMaxLoc(scores, 0, &conf, 0, &maxLoc);
+
+                conf = (yoloVersion != YoloVersion::YOLOv8) ? conf * obj_conf : conf;
+                if (conf < confThreshold)
+                    continue;
+
+                // get bbox coords
+                float* det = preds.ptr<float>(i);
+                double cx = det[0];
+                double cy = det[1];
+                double w = det[2];
+                double h = det[3];
+
+                // [x1, y1, x2, y2]
+                double x = cx - 0.5 * w;
+                double y = cy - 0.5 * h;
+                double width = darknet ? w : cx + 0.5 * w;
+                double height = darknet ? h : cy + 0.5 * h;
+
+                boxes.emplace_back(Rect2d(x, y, width, height));
+                classIds.emplace_back(maxLoc.x);
+                confidences.emplace_back(conf);
+            }
+        }
+
+        // NMS
+        std::vector<int> keep_idx;
+        NMSBoxes(boxes, confidences, confThreshold, nmsThreshold, keep_idx);
```

Comment:
NMS should be done within each class, or you lose overlapped boxes but actually they are from different classes.  Take this link for reference: https://github.com/opencv/opencv_zoo/blob/b6e370b10f641879a87890d44e42173077154a05/models/object_detection_yolox/yolox.py#L45-L69

Reply:
There is even a flag `nmsAcrossClasses` in `DetectionModel`

Reply:
Done!

Reply:
Thank you, but you can call `NMSBoxesBatched` to keep code clean and shorter.